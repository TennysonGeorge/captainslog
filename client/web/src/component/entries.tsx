import * as React from "react"
import { useEffect, useState } from "react"
import { KeyboardEvent } from "react"

import history from "../history"

import { Book, EntriesCreateRequest, Entry, EntryUnsaved } from "../definitions"
import { cachedGetBook, createEntries, getEntriesForBook } from "../remote"

import { DatePicker, Grouping } from "./date_picker"

const KEY_ENTER = 13

type EntriesGenerationId = {
  entries: EntryUnsaved[]
  createdAt: Date
}

const buildEntriesCreateRequest =
  (text: string, createdAt: Date): EntryUnsaved[] =>
    text.trim().split("\n").reduce((acc, line) => {
      if (line[0] === "#") {
        acc.createdAt = new Date(Date.parse(line.replace(/^#+/, "")))
      } else if (line) {
        acc.entries.push({
          createdAt: acc.createdAt.toISOString(),
          text: line,
        })
      }
      return acc
    }, { entries: [], createdAt } as EntriesGenerationId).entries

const noEntriesMessage = (date: Date, grouping: Grouping) => {
  switch (grouping) {
    case Grouping.NONE: return "No entries were found."
    case Grouping.DAY: return `No entries were found on ${date.toDateString()}.`
    default: return
  }
}

const genDatePicker = (date: Date, book: Book | null) =>
  !book || book.grouping === Grouping.NONE ? null :
    <DatePicker
      grouping={book.grouping}
      date={date}
      onChange={(d) => history.replace(`/book/${book.guid}/${+d}`)}
    />

type EntriesProps = {
  bookGuid: string
  date: Date
}

export const Entries = ({ bookGuid, date }: EntriesProps) => {
  const [text, setText] = useState("")
  const [book, setBook] = useState<Book | null>(null)
  const [entries, setEntries] = useState<Entry[]>([])

  const fetchEntries = () =>
    getEntriesForBook(bookGuid, date).then(setEntries)

  useEffect(() => {
    setEntries([])
    cachedGetBook(bookGuid)
      .then(setBook)
      .then(fetchEntries)
  }, [bookGuid, date])

  const handleKeyPress = (ev: KeyboardEvent<HTMLTextAreaElement>) => {
    if (ev.charCode !== KEY_ENTER || !book || !text.trim()) {
      return
    }

    createEntries(book.guid, buildEntriesCreateRequest(text, date))
      .then(fetchEntries)

    setText("")
    ev.preventDefault()
  }

  const dateInput = genDatePicker(date, book)
  const textInput =
    <textarea
      rows={1}
      placeholder="Enter a new log!"
      value={text}
      onChange={(e) => setText(e.target.value)}
      onKeyPress={handleKeyPress}
      className="entries-textarea"
    />

  const title = book ? <h1>{book.name}</h1> : null
  const header = !book ? null :
    <div className="entries-action-header">
      <div className="entries-action-header-col">{textInput}</div>
      <div className="entries-action-header-col">{dateInput}</div>
    </div>

  let content
  if (entries.length) {
    content = <EntryList items={entries} />
  } else if (book) {
    content = <div className="entries-empty">{noEntriesMessage(date, book.grouping)}</div>
  } else {
    content = null
  }

  return <div className="entries">
    {title}
    {header}
    {content}
  </div>
}

type HasCreatedAt = Pick<Entry, "createdAt">

const sortByCreatedAt = (a: HasCreatedAt, b: HasCreatedAt) => {
  if (a.createdAt === b.createdAt) {
    return 0
  } else if (a.createdAt > b.createdAt) {
    return -1
  } else {
    return 1
  }
}

const mmap = <V, R>(xs: { [index: string]: V }, fn: (k: string, v: V, i: number) => R) =>
  Object.keys(xs).map((k, i) => fn(k, xs[k], i))

type EntryListProps = {
  items: Entry[]
}

export const EntryList = ({ items }: EntryListProps) => {
  const sorted = items.sort(sortByCreatedAt)
  const elems = sorted.map((entry, i) => {
    const datalist = !entry.data ? null : mmap(entry.data, (key, val, j) =>
      <span key={j} className="entry-data-item">{key}: {val}</span>)

    return (
      <div key={entry.guid} className="entry">
        <div>{entry.text}</div>
        <div className="entry-data">{datalist}</div>
      </div>
    )
  })

  return <div>{elems}</div>
}
