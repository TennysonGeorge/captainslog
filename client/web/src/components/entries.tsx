import * as React from "react"
import { useEffect, useState } from "react"
import { KeyboardEvent } from "react"

import history from "../history"

import { Book, Entry, EntryCreateRequest } from "../definitions"
import { cachedGetBook, createEntry, getEntriesForBook } from "../remote"

import { DatePicker, Grouping } from "./date_picker"

const KEY_ENTER = 13

const buildEntryCreateRequest =
  (book: Book, text: string, createdAt: Date): EntryCreateRequest =>
    ({
      bookGuid: book.guid,
      createdAt: createdAt.toISOString(),
      guid: Math.random().toString(),
      text,
    })

const genDatePicker = (date: Date, book: Book | null) =>
  !book || book.grouping === Grouping.NONE ? null :
    <div>
      <DatePicker
        grouping={book.grouping}
        date={date}
        onChange={(d) => history.replace(`/${book.guid}/${+d}`)}
      />
    </div>

type EntryListViewProps = {
  date: Date
  bookGuid: string
}

export function EntryListView(props: EntryListViewProps) {
  const [text, setText] = useState("")
  const [book, setBook] = useState<Book | null>(null)
  const [entries, setEntries] = useState<Entry[]>([])

  const fetchEntries = () =>
    getEntriesForBook(props.bookGuid, props.date).then(setEntries)

  useEffect(() => {
    cachedGetBook(props.bookGuid)
      .then(setBook)
      .then(fetchEntries)
  }, [props.bookGuid, props.date])

  const handleKeyPress = (ev: KeyboardEvent<HTMLTextAreaElement>) => {
    if (ev.charCode !== KEY_ENTER || !book || !text.trim()) {
      return
    }

    createEntry(buildEntryCreateRequest(book, text.trim(), props.date))
      .then(fetchEntries)

    setText("")
    ev.preventDefault()
  }

  return <div>
    <textarea
      rows={1}
      placeholder="Enter a new log!"
      value={text}
      onChange={(e) => setText(e.target.value)}
      onKeyPress={handleKeyPress}
    />
    {genDatePicker(props.date, book)}
    <EntryList items={entries} />
  </div>
}

type HasCreatedAt = Pick<Entry, "createdAt">

const sortByCreatedAt = (a: HasCreatedAt, b: HasCreatedAt) => {
  if (a.createdAt === b.createdAt) {
    return 0
  } else if (a.createdAt > b.createdAt) {
    return -1
  } else {
    return 1
  }
}

const mmap = <V, R>(xs: { [index: string]: V }, fn: (k: string, v: V, i: number) => R) =>
  Object.keys(xs).map((k, i) => fn(k, xs[k], i))

type EntryListProps = {
  items: Entry[]
}

export const EntryList = ({ items }: EntryListProps) => {
  const sorted = items.sort(sortByCreatedAt)
  const elems = sorted.map((entry, i) => {
    const datalist = !entry.data ? null : mmap(entry.data, (key, val, j) =>
      <span key={j}>{key}: {val}</span>)

    return (
      <div key={entry.guid}>
        <div>{entry.text}</div>
        <div>{datalist}</div>
      </div>
    )
  })

  return <div>{elems}</div>
}
