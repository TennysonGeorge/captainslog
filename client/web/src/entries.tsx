import * as React from "react"
import { useEffect, useState } from "react"
import { KeyboardEvent } from "react"

import history from "./history"

import { Book, Entry, EntryCreateRequest, EntriesCreateRequest } from "./definitions"
import { cachedGetBook, createEntries, getEntriesForBook } from "./remote"

import { DatePicker, Grouping } from "./date_picker"

const KEY_ENTER = 13

type EntriesGenerationId = {
  entries: EntryCreateRequest[]
  createdAt: Date
}

const buildEntriesCreateRequest =
  (book: Book, text: string, createdAt: Date): EntriesCreateRequest =>
    text.trim().split("\n").reduce((acc, line) => {
      if (line[0] === "#") {
        acc.createdAt = new Date(Date.parse(line.replace(/^#+/, "")))
      } else if (line) {
        acc.entries.push({
          bookGuid: book.guid,
          createdAt: acc.createdAt.toISOString(),
          guid: Math.random().toString(),
          text: line,
        })
      }
      return acc
    }, { entries: [], createdAt } as EntriesGenerationId)

const genDatePicker = (date: Date, book: Book | null) =>
  !book || book.grouping === Grouping.NONE ? null :
    <DatePicker
      grouping={book.grouping}
      date={date}
      onChange={(d) => history.replace(`/${book.guid}/${+d}`)}
    />

type EntriesProps = {
  date: Date
  bookGuid: string
}

export const Entries = (props: EntriesProps) => {
  const [text, setText] = useState("")
  const [book, setBook] = useState<Book | null>(null)
  const [entries, setEntries] = useState<Entry[]>([])

  const fetchEntries = () =>
    getEntriesForBook(props.bookGuid, props.date).then(setEntries)

  useEffect(() => {
    setEntries([])
    cachedGetBook(props.bookGuid)
      .then(setBook)
      .then(fetchEntries)
  }, [props.bookGuid, props.date])

  const handleKeyPress = (ev: KeyboardEvent<HTMLTextAreaElement>) => {
    if (ev.charCode !== KEY_ENTER || !book || !text.trim()) {
      return
    }

    createEntries(buildEntriesCreateRequest(book, text, props.date))
      .then(fetchEntries)

    setText("")
    ev.preventDefault()
  }

  return <div className="entries">
    <h1>{book ? book.name : "\u00A0"}</h1>
    <div className="entries-action-header">
      <div className="entries-action-header-col">
        <textarea
          rows={1}
          placeholder="Enter a new log!"
          value={text}
          onChange={(e) => setText(e.target.value)}
          onKeyPress={handleKeyPress}
          className="entries-textarea"
        />
      </div>
      <div className="entries-action-header-col">
        {genDatePicker(props.date, book)}
      </div>
    </div>
    <EntryList items={entries} />
  </div>
}

type HasCreatedAt = Pick<Entry, "createdAt">

const sortByCreatedAt = (a: HasCreatedAt, b: HasCreatedAt) => {
  if (a.createdAt === b.createdAt) {
    return 0
  } else if (a.createdAt > b.createdAt) {
    return -1
  } else {
    return 1
  }
}

const mmap = <V, R>(xs: { [index: string]: V }, fn: (k: string, v: V, i: number) => R) =>
  Object.keys(xs).map((k, i) => fn(k, xs[k], i))

type EntryListProps = {
  items: Entry[]
}

export const EntryList = ({ items }: EntryListProps) => {
  const sorted = items.sort(sortByCreatedAt)
  const elems = sorted.map((entry, i) => {
    const datalist = !entry.data ? null : mmap(entry.data, (key, val, j) =>
      <span key={j} className="entry-data-item">{key}: {val}</span>)

    return (
      <div key={entry.guid} className="entry">
        <div>{entry.text}</div>
        <div className="entry-data">{datalist}</div>
      </div>
    )
  })

  return <div>{elems}</div>
}
