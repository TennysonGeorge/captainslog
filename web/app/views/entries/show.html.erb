<%= render "layouts/header" %>

<div class="content-view">
  <h1><%= t(:edit_entry) %></h1>
  <%= render "entry/edit_form", :entry => entry %>
</div>
