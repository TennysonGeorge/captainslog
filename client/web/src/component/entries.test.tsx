import * as React from "react"
import { act } from "react-dom/test-utils"
import { mount } from "enzyme"

import "./testsetup"

import { EntryList } from "./entries"

test("snapshot EntryList", () => {
  const day = new Date(1566096667888).toString()
  const entries = [
    { guid: "e1", text: "entry #1", createdAt: day, updatedAt: day, data: { a: "A1", b: "B1", c: "C1" } },
    { guid: "e2", text: "entry #2", createdAt: day, updatedAt: day, data: { a: "A2", b: "B2", c: "C2" } },
    { guid: "e3", text: "entry #3", createdAt: day, updatedAt: day, data: { a: "A3", b: "B3", c: "C3" } },
  ]
  const component = mount(<EntryList items={entries} />)
  expect(component).toMatchSnapshot()
})
