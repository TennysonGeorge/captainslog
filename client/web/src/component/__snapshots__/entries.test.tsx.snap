// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot EntryList 1`] = `
<Component
  items={
    Array [
      Object {
        "createdAt": "Sat Aug 17 2019 20:51:07 GMT-0600 (Mountain Daylight Time)",
        "data": Object {
          "a": "A1",
          "b": "B1",
          "c": "C1",
        },
        "guid": "e1",
        "text": "entry #1",
        "updatedAt": "Sat Aug 17 2019 20:51:07 GMT-0600 (Mountain Daylight Time)",
      },
      Object {
        "createdAt": "Sat Aug 17 2019 20:51:07 GMT-0600 (Mountain Daylight Time)",
        "data": Object {
          "a": "A2",
          "b": "B2",
          "c": "C2",
        },
        "guid": "e2",
        "text": "entry #2",
        "updatedAt": "Sat Aug 17 2019 20:51:07 GMT-0600 (Mountain Daylight Time)",
      },
      Object {
        "createdAt": "Sat Aug 17 2019 20:51:07 GMT-0600 (Mountain Daylight Time)",
        "data": Object {
          "a": "A3",
          "b": "B3",
          "c": "C3",
        },
        "guid": "e3",
        "text": "entry #3",
        "updatedAt": "Sat Aug 17 2019 20:51:07 GMT-0600 (Mountain Daylight Time)",
      },
    ]
  }
>
  <div>
    <div
      className="entry"
      key="e1"
    >
      <div>
        entry #1
      </div>
      <div
        className="entry-data"
      >
        <span
          className="entry-data-item"
          key="0"
        >
          a
          : 
          A1
        </span>
        <span
          className="entry-data-item"
          key="1"
        >
          b
          : 
          B1
        </span>
        <span
          className="entry-data-item"
          key="2"
        >
          c
          : 
          C1
        </span>
      </div>
    </div>
    <div
      className="entry"
      key="e2"
    >
      <div>
        entry #2
      </div>
      <div
        className="entry-data"
      >
        <span
          className="entry-data-item"
          key="0"
        >
          a
          : 
          A2
        </span>
        <span
          className="entry-data-item"
          key="1"
        >
          b
          : 
          B2
        </span>
        <span
          className="entry-data-item"
          key="2"
        >
          c
          : 
          C2
        </span>
      </div>
    </div>
    <div
      className="entry"
      key="e3"
    >
      <div>
        entry #3
      </div>
      <div
        className="entry-data"
      >
        <span
          className="entry-data-item"
          key="0"
        >
          a
          : 
          A3
        </span>
        <span
          className="entry-data-item"
          key="1"
        >
          b
          : 
          B3
        </span>
        <span
          className="entry-data-item"
          key="2"
        >
          c
          : 
          C3
        </span>
      </div>
    </div>
  </div>
</Component>
`;
